src/query_graph/graph_path.rs

```diff
@@ -0,0 +1,228 @@
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct SubgraphEnteringEdgeInfo {
+    /// The index within the `edges` array.
+    index: usize,
+    /// The edge within the `edges` array (aliased here for convenience).
+    edge: EdgeIndex,
+    /// The cost of resolving the conditions for this edge.
+    conditions_cost: QueryPlanCost,
+}
+
+/// A `GraphPath` whose triggers are operation elements (essentially meaning that the path has been
+/// guided by a GraphQL operation).
+// PORT_NOTE: As noted in the docs for `GraphPath`, we omit a type parameter for the root node,
+// whose constraint is instead checked at runtime. This means the `OpRootPath` type in the JS
+// codebase is replaced with this one.
+pub(crate) type OpGraphPath = GraphPath<OpGraphPathTrigger, Option<EdgeIndex>>;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) enum OpGraphPathTrigger {
+    Field(OpPathField),
+    InlineFragment(OpPathInlineFragment),
+    Context(OpGraphPathContext),
+}
+
+/// A path of operation elements within a GraphQL operation.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) struct OpPath(Vec<Arc<OpPathElement>>);
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) enum OpPathElement {
+    Field(OpPathField),
+    InlineFragment(OpPathInlineFragment),
+}
+
+/// A field within a GraphQL operation, minus the selection set.
+#[derive(Debug, Clone)]
+pub(crate) struct OpPathField {
+    schema: Arc<ValidFederationSchema>,
+    position: FieldDefinitionPosition,
+    alias: Option<Name>,
+    arguments: Vec<Node<Argument>>,
+    directives: DirectiveList,
+}
+
+impl Eq for OpPathField {}
+
+impl PartialEq for OpPathField {
+    fn eq(&self, _other: &Self) -> bool {
+        todo!()
+    }
+}
+
+impl Hash for OpPathField {
+    fn hash<H: Hasher>(&self, _state: &mut H) {
+        todo!()
+    }
+}
+
+/// An inline fragment within a GraphQL operation, minus the selection set.
+#[derive(Debug, Clone)]
+pub(crate) struct OpPathInlineFragment {
+    schema: Arc<ValidFederationSchema>,
+    parent_type_position: CompositeTypeDefinitionPosition,
+    type_condition: Option<CompositeTypeDefinitionPosition>,
+    directives: DirectiveList,
+}
+
+impl Eq for OpPathInlineFragment {}
+
+impl PartialEq for OpPathInlineFragment {
+    fn eq(&self, _other: &Self) -> bool {
+        todo!()
+    }
+}
+
+impl Hash for OpPathInlineFragment {
+    fn hash<H: Hasher>(&self, _state: &mut H) {
+        todo!()
+    }
+}
+
+/// Records, as we walk a path within a GraphQL operation, important directives encountered
+/// (currently `@include` and `@skip` with their conditions).
+#[derive(Debug, Clone, Eq)]
+pub(crate) struct OpGraphPathContext {
+    /// A list of conditionals (e.g. `[{ kind: Include, value: true}, { kind: Skip, value: $foo }]`)
+    /// in the reverse order in which they were applied (so the first element is the inner-most
+    /// applied include/skip).
+    conditionals: Vec<Arc<OperationConditional>>,
```

Comment:
Can we ever have more than two of these? Or could the Vec be replaced with `include: Option<BooleanOrVariable>, skip: Option<BooleanOrVariable>`?

Reply:
Yeah, there can be a variable number (the example just happens to have two).